### PR TITLE
[lldb] Adjust DynamicLoaderDarwin::GetThreadLocalData to support changes to TLS on macOS

### PR DESCRIPTION
**Pull request by bulbazord**
TLS implementation on Apple OSes has changed. Instead of acquiring a pthread_key and calling pthread_getspecific, we instead acquire a function pointer and perform a function call with it. This fixes accessing thread local storage on macOS 14 (Sonoma) and newer.

Note: Some versions of Apple's new linker do not emit debug symbols for TLS symbols. This causes the TLS tests to fail because LLDB and dsymutil expects there to be debug symbols to resolve the relevant TLS block. You may work around this by switching to the older linker or disabling the test until you have a newer version of the new linker.

rdar://120676969

### CODE REVIEW

File: lldb/source/Plugins/DynamicLoader/MacOSX-DYLD/DynamicLoaderDarwin.cpp [hunk 1]
```diff
@@ -1048,70 +1048,107 @@ DynamicLoaderDarwin::GetThreadLocalData(const lldb::ModuleSP module_sp,
...
+  // approximately like this:
+  // struct TLS_Thunk {
+  //  void *(*get_addr)(struct TLS_Thunk *);
+  //  size_t key;
+  //  size_t offset;
+  // }
+  //
+  // The strategy is to take get_addr from the structure, call it with the
+  // address of the containing TLS_Thunk structure, and add the offset to the
+  // resulting pointer.
```
Comment (bulbazord) [inline lldb/source/Plugins/DynamicLoader/MacOSX-DYLD/DynamicLoaderDarwin.cpp:1091]: I need to double check this actually. Not sure if we need to add the offset.

File: lldb/source/Plugins/DynamicLoader/MacOSX-DYLD/DynamicLoaderDarwin.cpp [hunk 2]
```diff
@@ -1048,70 +1048,107 @@ DynamicLoaderDarwin::GetThreadLocalData(const lldb::ModuleSP module_sp,
...
+  DiagnosticManager execution_errors;
+  ExecutionContext exe_ctx(thread_sp);
+
+  // On modern apple platforms, there is a small data structure that looks
+  // approximately like this:
+  // struct TLS_Thunk {
+  //  void *(*get_addr)(struct TLS_Thunk *);
+  //  size_t key;
+  //  size_t offset;
+  // }
```
Comment (felipepiovezan) [inline lldb/source/Plugins/DynamicLoader/MacOSX-DYLD/DynamicLoaderDarwin.cpp:1087]: Do you think this comment would be more appropriate a few lines above? It seems like we are trying to find these fields around line 1067

Comment (JDevlieghere) [reply lldb/source/Plugins/DynamicLoader/MacOSX-DYLD/DynamicLoaderDarwin.cpp:1087]: +1 as this is also shared between the "old" and "new" implementation.

**Review by JDevlieghere — COMMENTED** [review]
Currently this is a little hard to read given a bunch of shared setup followed by two possible ways to read the TLS data. If those implementations don't require too many inputs, it might be worth splitting this out into two helper functions to make this easier to follow and extend in the future.

**Comment by bulbazord** [comment]
> Currently this is a little hard to read given a bunch of shared setup followed by two possible ways to read the TLS data. If those implementations don't require too many inputs, it might be worth splitting this out into two helper functions to make this easier to follow and extend in the future.

I can do that. Since this change is quite large as-is, perhaps I should break this up into smaller changes and submit them separately?

**Review by JDevlieghere — APPROVED** [review]

**Comment by clayborg** [comment]
It will be nice to not have the TLS test failing on macOS anymore!

**Comment by bulbazord** [comment]
I broke up this PR into multiple commits and resubmitted at https://github.com/llvm/llvm-project/pull/77988. Please take a look!

**Pull request closed without merging**